### PR TITLE
Add code refinement

### PR DESCRIPTION
**Pull request by reshinthadithyan**
### OVERVIEW
Millions of open-source projects with numerous bug fixes
are available in code repositories. This proliferation
of software development histories can be leveraged to
learn how to fix common programming bugs
Code refinement aims to automatically fix bugs in the code,
which can contribute to reducing the cost of bug-fixes for developers.
Given a piece of Java code with bugs,
the task is to remove the bugs to output the refined code.

### CODE REVIEW

File: datasets/code_refinement/code_refinement.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+                    "fix": datasets.Value("string"),
+                }
+            )
+        else:  # This is an example to show how to have different features for "first_domain" and "second_domain"
+            features = datasets.Features(
+                {
+                    "bug": datasets.Value("string"),
+                    "fix": datasets.Value("string"),
+                }
+            )
```
Comment (lhoestq) [inline datasets/code_refinement/code_refinement.py:117]:
```suggestion
        features = datasets.Features(
            {
                "bug": datasets.Value("string"),
                "fix": datasets.Value("string"),
            }
        )
```
They're the same, no need to check for the config name

File: datasets/code_refinement/code_refinement.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        return datasets.DatasetInfo(
+            description=_DESCRIPTION,
+            features=features,  # Here we define them above because they are different between the two configurations
+            supervised_keys=["code", "ml4code"],
```
Comment (lhoestq) [inline datasets/code_refinement/code_refinement.py:121]:
```suggestion
            supervised_keys=("bug", "fix"),
```

**Review by lhoestq — COMMENTED** [review]
Looking pretty good so far ! thanks

Could you add the dataset tags ? They're the yaml part at the top of the dataset card.
see more info here: https://github.com/huggingface/datasets/blob/master/ADD_NEW_DATASET.md#tag-the-dataset-and-write-the-dataset-card

File: datasets/code_refinement/code_refinement.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        "test_fix": "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/test.buggy-fixed.fixed",
+        "train_bug": "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/train.buggy-fixed.buggy",
+        "valid_bug": "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/valid.buggy-fixed.buggy",
+        "test_bug": "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/test.buggy-fixed.buggy",
+    },
+}
+
+
+# TODO: Name of the dataset usually match the script name with CamelCase instead of snake_case
+class NewDataset(datasets.GeneratorBasedBuilder):
```
Comment (lhoestq) [inline datasets/code_refinement/code_refinement.py:71]:
```suggestion
class CodeRefinement(datasets.GeneratorBasedBuilder):
```
this needs to be the camel case version of the dataset script name `code_refinement`

File: datasets/code_refinement/dataset_infos.json [hunk 4]
```diff
@@ -0,0 +1 @@
...GLUE/main/Code-Code/code-refinement/data/medium/valid.buggy-fixed.fixed": {"num_bytes": 1979872, "checksum": "7ef5e4b2e95914e0eceb4f2cf6dfad0641625145319b9836db70d3f8745ad2d6"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/test.buggy-fixed.fixed": {"num_bytes": 1970531, "checksum": "4b13298647e9a782bf908d4a26710e97a1846f5513a9bf1aa46ac8223fb84b3d"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/train.buggy-fixed.buggy": {"num_bytes": 16188348, "checksum": "4570731680fa183650864e8729a7354d235c9a3ef42f0085ace3441418074085"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/valid.buggy-fixed.buggy": {"num_bytes": 2028309, "checksum": "8ad01f88be2009599007f40427458d6d2601fe93f2f1d65b0f46b7d414a3add2"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/test.buggy-fixed.buggy": {"num_bytes": 2014594, "checksum": "21107528c3b25bfdec24d0c4c18a953de31c26f3795a7d7c9e108a60396bcd38"}}, "download_size": 39979724, "post_processing_size": null, "dataset_size": 40503428, "size_in_bytes": 80483152}}
```
Comment (lhoestq) [inline datasets/code_refinement/dataset_infos.json:1]:
```suggestion
{"code_refinement_small": {"description": "Millions of open-source projects with numerous bug fixes\nare available in code repositories. This proliferation\nof software development histories can be leveraged to\nlearn how to fix common programming bugs\nCode refinement aims to automatically fix bugs in the code,\nwhich can contribute to reducing the cost of bug-fixes for developers.\nGiven a piece of Java code with bugs,\nthe task is to remove the bugs to output the refined code. ", "citation": "@article{CodeXGLUE,\n  title={CodeXGLUE: A Benchmark Dataset and Open Challenge for Code Intelligence},\n  year={2020},\n}\n", "homepage": "https://github.com/microsoft/CodeXGLUE", "license": "MIT License", "features": {"bug": {"dtype": "string", "id": null, "_type": "Value"}, "fix": {"dtype": "string", "id": null, "_type": "Value"}}, "post_processed": null, "supervised_keys": {"input": "code", "output": "ml4code"}, "builder_name": "code_refinement", "config_name": "code_refinement_small", "version": {"version_str": "0.0.0", "description": null, "major": 0, "minor": 0, "patch": 0}, "splits": {"train": {"name": "train", "num_bytes": 12819999, "num_examples": 46680, "dataset_name": "code_refinement"}, "validation": {"name": "validation", "num_bytes": 1605910, "num_examples": 5835, "dataset_name": "code_refinement"}, "test": {"name": "test", "num_bytes": 1596368, "num_examples": 5835, "dataset_name": "code_refinement"}}, "download_checksums": {"https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/small/train.buggy-fixed.fixed": {"num_bytes": 5936570, "checksum": "c98b1139265d33e787a9dd742a464e7eb5bd137ebb3fcb54f0416ee7672739f3"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/small/valid.buggy-fixed.fixed": {"num_bytes": 743907, "checksum": "dc9c23594350988dcdaf456a1e8eea1dc86c81d1c0ff9985ff6d63d15196ffd2"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/small/test.buggy-fixed.fixed": {"num_bytes": 739739, "checksum": "e7ec462d00d253ddec3d1cef7a06cfc99db1cfe94039a6607ed14afeef6be04a"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/small/train.buggy-fixed.buggy": {"num_bytes": 6509949, "checksum": "dfb4366dedb73dd40f78c3af870ccb0a1aeff2d9ceb45585df26c99897740748"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/small/valid.buggy-fixed.buggy": {"num_bytes": 815315, "checksum": "e83a0c524cdce5a4492dfe0c9bc7d642aa5ed267ddbae5ffd1e563fe54cae6b8"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/small/test.buggy-fixed.buggy": {"num_bytes": 809941, "checksum": "d2e675094f471b3bfbb9419eb7cf14dca11f8d41185319d3310cbc3f517df323"}}, "download_size": 15555421, "post_processing_size": null, "dataset_size": 16022277, "size_in_bytes": 31577698}, "code_refinement_medium": {"description": "Millions of open-source projects with numerous bug fixes\nare available in code repositories. This proliferation\nof software development histories can be leveraged to\nlearn how to fix common programming bugs\nCode refinement aims to automatically fix bugs in the code,\nwhich can contribute to reducing the cost of bug-fixes for developers.\nGiven a piece of Java code with bugs,\nthe task is to remove the bugs to output the refined code. ", "citation": "@article{CodeXGLUE,\n  title={CodeXGLUE: A Benchmark Dataset and Open Challenge for Code Intelligence},\n  year={2020},\n}\n", "homepage": "https://github.com/microsoft/CodeXGLUE", "license": "MIT License", "features": {"bug": {"dtype": "string", "id": null, "_type": "Value"}, "fix": {"dtype": "string", "id": null, "_type": "Value"}}, "post_processed": null, "supervised_keys": {"input": "code", "output": "ml4code"}, "builder_name": "code_refinement", "config_name": "code_refinement_medium", "version": "0.0.0", "splits": {"train": {"name": "train", "num_bytes": 32405378, "num_examples": 52364, "dataset_name": "code_refinement"}, "validation": {"name": "validation", "num_bytes": 4060557, "num_examples": 6546, "dataset_name": "code_refinement"}, "test": {"name": "test", "num_bytes": 4037493, "num_examples": 6545, "dataset_name": "code_refinement"}}, "download_checksums": {"https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/train.buggy-fixed.fixed": {"num_bytes": 15798070, "checksum": "009c121662602642bc55f6882f220aea6a738e6a11f2c4df86e7fe3cd30c175c"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/valid.buggy-fixed.fixed": {"num_bytes": 1979872, "checksum": "7ef5e4b2e95914e0eceb4f2cf6dfad0641625145319b9836db70d3f8745ad2d6"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/test.buggy-fixed.fixed": {"num_bytes": 1970531, "checksum": "4b13298647e9a782bf908d4a26710e97a1846f5513a9bf1aa46ac8223fb84b3d"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/train.buggy-fixed.buggy": {"num_bytes": 16188348, "checksum": "4570731680fa183650864e8729a7354d235c9a3ef42f0085ace3441418074085"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/valid.buggy-fixed.buggy": {"num_bytes": 2028309, "checksum": "8ad01f88be2009599007f40427458d6d2601fe93f2f1d65b0f46b7d414a3add2"}, "https://raw.githubusercontent.com/microsoft/CodeXGLUE/main/Code-Code/code-refinement/data/medium/test.buggy-fixed.buggy": {"num_bytes": 2014594, "checksum": "21107528c3b25bfdec24d0c4c18a953de31c26f3795a7d7c9e108a60396bcd38"}}, "download_size": 39979724, "post_processing_size": null, "dataset_size": 40503428, "size_in_bytes": 80483152}}
```

**Review by lhoestq — COMMENTED** [review]
I fixed the class name

**Comment by lhoestq** [comment]
Also cc @madlag since I recall you wanted to work on CodeXGlue as well ?

**Comment by madlag** [comment]
Yes, sorry I did not see earlier your message. I added 34 on the 35 datasets in CodeXGlue, tomorrow I will wrap it up, and so I will remove my version for code_refinement. Maybe we can just have a renaming after the merge, to have a consistent naming with all the other codexglue datasets ? What do you think @reshinthadithyan ?

**Comment by reshinthadithyan** [comment]
> Yes, sorry I did not see earlier your message. I added 34 on the 35 datasets in CodeXGlue, tomorrow I will wrap it up, and so I will remove my version for code_refinement. Maybe we can just have a renaming after the merge, to have a consistent naming with all the other codexglue datasets ? What do you think @reshinthadithyan ?

Hello @madlag, I think you can retain that in your script. Let's stick onto the same file like how Glue is maintained.

**Comment by yjernite** [comment]
Hi @reshinthadithyan ! Are you still working on this version of the dataset or are we going with @madlag 's only ?

**Comment by reshinthadithyan** [comment]
> Hi @reshinthadithyan ! Are you still working on this version of the dataset or are we going with @madlag 's only ?

Hello, yes. We are going with Madlag's

**Pull request closed without merging**